tst/unit/test_mesh_data.cpp

```diff
@@ -0,0 +1,148 @@
+using parthenon::loop_pattern_mdrange_tag;
+using parthenon::MeshBlock;
+using parthenon::MeshBlockData;
+using parthenon::MeshData;
+using parthenon::Metadata;
+using parthenon::PackIndexMap;
+using parthenon::par_for;
+using parthenon::Real;
+using parthenon::StateDescriptor;
+
+BlockList_t MakeBlockList(const std::shared_ptr<StateDescriptor> pkg, const int NBLOCKS,
+                          const int NSIDE, const int NDIM) {
+  BlockList_t block_list;
+  block_list.reserve(NBLOCKS);
+  for (int i = 0; i < NBLOCKS; ++i) {
+    auto pmb = std::make_shared<MeshBlock>(NSIDE, NDIM);
+    auto &pmbd = pmb->meshblock_data.Get();
+    pmbd->Initialize(pkg, pmb);
+    block_list.push_back(pmb);
+  }
+  return block_list;
+}
+
+TEST_CASE("MeshData works as expected for simple packs", "[MeshData]") {
+  GIVEN("A set of meshblocks and meshblock and mesh data") {
+    constexpr int N = 6;
+    constexpr int NDIM = 3;
+    constexpr int NBLOCKS = 9;
+    std::vector<int> scalar_shape{N, N, N};
+    std::vector<int> vector_shape{N, N, N, 3};
+
+    Metadata m({Metadata::Independent, Metadata::WithFluxes}, scalar_shape);
+    Metadata m_vector({Metadata::Independent, Metadata::WithFluxes, Metadata::Vector},
+                      vector_shape);
+    auto pkg = std::make_shared<StateDescriptor>("Test package");
+    pkg->AddField("v1", m);
+    pkg->AddField("v3", m_vector);
+    pkg->AddField("v5", m);
+    BlockList_t block_list = MakeBlockList(pkg, NBLOCKS, N, NDIM);
+
+    MeshData<Real> mesh_data;
+    mesh_data.Set(block_list, "base");
+
+    THEN("The number of blocks is correct") { REQUIRE(mesh_data.NumBlocks() == NBLOCKS); }
+
+    WHEN("We initialize the independent variables by hand") {
+      auto ib = block_list[0]->cellbounds.GetBoundsI(IndexDomain::entire);
+      auto jb = block_list[0]->cellbounds.GetBoundsJ(IndexDomain::entire);
+      auto kb = block_list[0]->cellbounds.GetBoundsK(IndexDomain::entire);
+      std::vector<std::string> all_indep{"v1", "v3", "v5"};
+      for (int b = 0; b < NBLOCKS; ++b) {
+        auto &pmb = block_list[b];
+        auto &pmbd = pmb->meshblock_data.Get();
+        for (int v = 0; v < all_indep.size(); ++v) {
+          auto &vnam = all_indep[v];
+          auto var = pmbd->Get(vnam);
+          auto var4 = var.data.Get<4>();
+          int num_components = var.GetDim(4);
+          par_for(
+              loop_pattern_mdrange_tag, "initialize " + vnam, DevExecSpace(), kb.s, kb.e,
+              jb.s, jb.e, ib.s, ib.e, KOKKOS_LAMBDA(int k, int j, int i) {
+                for (int c = 0; c < num_components; ++c) {
+                  Real n = i + 1e1 * j + 1e2 * k + 1e3 * c + 1e4 * v + 1e5 * b;
+                  var4(c, k, j, i) = n;
+                }
+              });
+        }
+      }
+
+      THEN("A pack for a scalar with a PackIndexMap works") {
+        PackIndexMap imap;
+        std::vector<std::string> vlist{"v5"};
+        auto pack = mesh_data.PackVariables(vlist, imap);
```

Comment:
```suggestion
        const std::vector<std::string> vlist{"v5"};
        const auto &pack = mesh_data.PackVariables(vlist, imap);
```

Reply:
The string can be a const, but the pack cannot be a reference. Since the pack gets copied to device, this will cause memory errors.

Reply:
Really? I thought a copy will be made to transfer to the device because the lambda is capture by value...

Reply:
Capturing a reference by value copies the reference, right? Maybe I'm misunderstanding something.

Reply:
I'm actually not sure. I thought it would feed the const ref into the copy constructor to make a copy, but you may be right. I guess we have to test it.

Reply:
I think it makes a copy. This code
```c++
#include <cstdio>

int main(int, char**) {
  const int var = 3;
  const int &ref = var;

  printf("&var = %p, &ref = %p\n", &var, &ref);

  auto lam_ref = [&]() {
    printf("In lambda ref:\n");
    printf("&var = %p, &ref = %p\n", &var, &ref);
  };

  auto lam_val = [=]() {
    printf("In lambda val:\n");
    printf("&var = %p, &ref = %p\n", &var, &ref);
  };

  lam_ref();
  lam_val();

  return 0;
}
```
produces this output for me:
```
&var = 0x7ffffc182004, &ref = 0x7ffffc182004
In lambda ref:
&var = 0x7ffffc182004, &ref = 0x7ffffc182004
In lambda val:
&var = 0x7ffffc181fe8, &ref = 0x7ffffc181fec
```
So in the second lambda that captures by reference, both pointers are different, because it made two fresh copies.

Reply:
ah ok interesting. Thanks for checking that. I swear I've been burned by this when working with kokkos before.